### PR TITLE
(fix) startup handling and logging of docker images tweaked

### PR DESCRIPTION
**Pull request by tobitege**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

Startup handling and logging of docker images tweaked to be clearer to user.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

Docker image was potentially pulled even though it already existed locally.
Improved logging of steps so there is no impression of some process hanging during startup.

---
**Link of any specific issues this addresses**

Helps #3551  #3644

### CODE REVIEW

File: openhands/runtime/builder/docker.py [hunk 1]
```diff
@@ -61,22 +61,29 @@ def build(self, path: str, tags: list[str]) -> str:
...
+            logger.info(f'Checking, if image {image_name} exists locally.')
+            self.docker_client.images.get(image_name)
+            logger.info(f'Image {image_name} found locally.')
+            return True
+        except docker.errors.ImageNotFound:
+            try:
+                logger.info(
+                    'Image not found locally. Trying to pull it, please wait...'
+                )
+                self.docker_client.images.pull(image_name)
```
Comment (xingyaoww) [inline openhands/runtime/builder/docker.py:81]: nit: We could also print the pulling processing though

File: openhands/runtime/utils/runtime_build.py [hunk 2]
```diff
@@ -260,9 +262,9 @@ def build_runtime_image(
     # Scenario 2: If a Docker image with the exact hash is not found, we will FIRST try to re-build it
     # by leveraging the `generic_runtime_image_name` to save some time
     # from re-building the dependencies (e.g., poetry install, apt install)
-    elif runtime_builder.image_exists(generic_runtime_image_name) and not force_rebuild:
+    if not force_rebuild and runtime_builder.image_exists(generic_runtime_image_name):
```
Comment (xingyaoww) [inline openhands/runtime/utils/runtime_build.py:265]:
```suggestion
    elif not force_rebuild and runtime_builder.image_exists(generic_runtime_image_name):
```

**Review by xingyaoww — APPROVED** [review]
LGTM!

File: openhands/runtime/utils/runtime_build.py (hunk 2 above)
Comment (tobitege) [reply openhands/runtime/utils/runtime_build.py:265]: If the first IF results in true, it returns. I don't see why there should be an "elif"?

File: openhands/runtime/builder/docker.py (hunk 1 above)
Comment (tobitege) [reply openhands/runtime/builder/docker.py:81]: You mean download progress bar? 🤔

File: openhands/runtime/utils/runtime_build.py (hunk 2 above)
Comment (xingyaoww) [reply openhands/runtime/utils/runtime_build.py:265]: I think this is more like a style preference i have... to make things look more like a "switch". Though it is just nits, feel free to ignore it.

File: openhands/runtime/builder/docker.py (hunk 1 above)
Comment (xingyaoww) [reply openhands/runtime/builder/docker.py:81]: yeah, i think .pull might returns a logger of some sort, if so, we can just loop through and print?

Comment (tobitege) [reply openhands/runtime/builder/docker.py:81]: I'll look into that for a follow up PR.

**Comment by tobitege** [comment]
> LGTM!

Thanks for jumping in so fast! 🙏

**Pull request merged**